### ISSUE TITLE
Add new Date field widget to utilize HTML5 #date input type

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4255]
Follow-up issue for https://github.com/backdrop/backdrop-issues/issues/3972

Now that we have html5 date fields in core, we should also provide them in core date module.
Possibly even replace the jQuery-based datepicker completely (not sure yet, if that's actually possible).

At least it should be possible for admins / site builders to provide consistent date field widgets for node (entity) forms.

**Comment by herbdool** [comment]
Perhaps a first step would be to provide html5 date as another widget option (and make it default) and keep the jQuery datepicker around but deprecated?

**Comment by BWPanda** [comment]
Closed https://github.com/backdrop/backdrop-issues/issues/4733 as a duplicate. Copying my comment over:

> There exists an HTML input type: date which provides native support for date fields, with a popup calendar, for modern browsers and mobile devices: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date
However browser support is spotty, so we can't implement this as the default solution.
>
> I propose adding a new, core widget (but not set as default) for the Date field which uses this native format. Then people who want to support mobile and modern browsers can do so, while others can use the existing 'popup calendar' widget.
>
> A contrib solution is an option, considering lack of consistency with browser support and possible confusion with existing widgets...

**Comment by BWPanda** [comment]
Just noticed that the 'Authored on' node field uses `#date`. Makes me feel better about adding this straight to core (as opposed to contrib).

**Comment by indigoxela** [comment]
@BWPanda your current concerns have already been discussed in the previous issue. Consensus was that it's OK and that a new [field widget would make sense](https://github.com/backdrop/backdrop-issues/issues/3972#issuecomment-570086110).

**Comment by BWPanda** [comment]
I tried doing a PR for this myself, but found it too complicated.

**Comment by indigoxela** [comment]
> I tried doing a PR for this myself, but found it too complicated.

It requires descending to the dungeons of date module code... :fearful:

**Comment by yorkshire-pudding** [comment]
I would welcome this.  I have just raised #5703 related to the timepicker so Backdrop is screaming out for a modern, usable date picker.

**Comment by indigoxela** [comment]
It's not that nobody else wants proper html5 date and time inputs - many do. It's that the date granularity setting sort of prevents that. All core date fields have that odd granularity setting. And as it's a field setting, the widget can't do anything about it.

But there are more things to fight with, when touching date module code...

**Comment by yorkshire-pudding** [comment]
Thanks @indigoxela - I'd picked up on some of the complexities while reading this issue.  I don't expect a magic bullet, just wanted to add my voice as saying we _should_ have this

**Comment by indigoxela** [comment]
It might be (technically) possible for the `datestamp` type "Date (Unix timestamp)" - and then simply ignore the granularity. But then there are still several traps.

We probably need a shiny, brave knight to kill that beast. :laughing:

**Comment by indigoxela** [comment]
Here we go, a functioning widget: [PR](https://github.com/backdrop/backdrop/pull/4157) (the test failure is a random one).

This will still need work, but testing and feedback is already welcome!

Screenshots:

![example-date-1](https://user-images.githubusercontent.com/842066/183286187-46c818b8-a934-4a5a-abcd-5a94cf85d20b.png)

![example-date-2](https://user-images.githubusercontent.com/842066/183286189-b8663d47-b555-4f83-86c6-a827b50e0709.png)

I tried to avoid reusing any of the existing (and complex) widget code. Purely addition.

**Comment by herbdool** [comment]
Nice work @indigoxela. I haven't tested it but took a look at the code. I agree with approach of adding a new widget instead of replacing an existing one.

**Comment by docwilmot** [comment]
The UI is refreshingly nice.

**Comment by bugfolder** [comment]
Did some quick tests on the sandbox. Very nice!

**Comment by indigoxela** [comment]
As this widget is new, and the date input (form API) has the ability to limit by "max" and "min" values, we are now able to provide an easier way to limit dates in a smarter way.

Here's a quick mockup of what that could look like.

All collapsed if no limit:

![no-limit](https://user-images.githubusercontent.com/842066/183353670-dc83a5d0-549f-4ca0-840c-746c624dcbd4.png)

Number fields if the limit is "traditionally" by year:

![by-year](https://user-images.githubusercontent.com/842066/183353746-d3df2266-ca42-4732-8bd7-12d5f42fdf81.png)

Text fields for php patterns, if the limit should be based on relative dates:

![bydate](https://user-images.githubusercontent.com/842066/183353809-41760600-701c-411d-bae8-00cb5994c420.png)

Do we want that? All of it or parts? Wording?

**Comment by yorkshire-pudding** [comment]
Thank you @indigoxela - this (the widget) is brilliant - it works great and would be a fantastic addition.

With regard to the limiting, I think the mockups make sense and it would be a valuable addition, but I'm concerned that including it within this issue and PR could complicate matters and slow down implementation of the widget.  

It would be great if we could get the widget into 1.23.  Perhaps the 'limiting' addition should be a separate issue and if that can also get in, great, but if not we would still get the widget?

**Comment by indigoxela** [comment]
> ... but I'm concerned that including it within this issue and PR could complicate matters and slow down implementation of the widget.

Good point. We could defer the "By date" limit to a follow up. But I'd still recommend to decide the "Radios" usage early, so we don't need to consider bigger changes (possibly update hooks), when we want to add new features to the limits.
I'm open to suggestions, though.

**Comment by yorkshire-pudding** [comment]
> This will still need work, but testing and feedback is already welcome!

Switching to this widget on an existing field that currently uses pop-up calendar causes a slightly odd reaction; it locks the date to the current year (I've tried with a couple of pre-existing year limit conditions).  If you then do configure and save it unlocks it.   Not a problem to reset, but may be unexpected behaviour and users may not remember what it was previously set to.  I think this might be what you were referring to when you said needs work.

**Comment by indigoxela** [comment]
> Switching to this widget on an existing field that currently uses pop-up calendar ... it locks the date to the current year...

Strange - can not reproduce it (yet). Maybe I misunderstand what "it locks the date" means... The limit on years or the limit on actual dates on node forms?

**Comment by yorkshire-pudding** [comment]
I'll try in a fresh lando app shortly in case it's interacting with something I've done in that particular instance.

**Comment by indigoxela** [comment]
I think, I found something, that might be related - not sure. Can you update to the latest commit? And many thanks for testing!

> this might be what you were referring to when you said needs work.

Not specifically - it's just that this thing is completely new and was written in relatively short time - it MUST be buggy somewhere. :laughing:

**Comment by yorkshire-pudding** [comment]
@indigoxela - that latest commit works to remove the restriction. 
However, unlike the transition between select list and pop-up calendar (and vice versa) (NB - text field has no year validation), the transition to (and from - though why anyone would want to move from it is a mystery) the HTML5 widget does not inherit the settings of the previous widget as its default.

Not an issue, but neither of the current widgets have a no limits option.
I wonder if Backdrop should have a hard limit aligned with the introduction of the Gregorian calendar 😆

**Comment by indigoxela** [comment]
> ... the HTML5 widget does not inherit the settings of the previous widget as its default.

Ah, I think now I get it. It can't "inherit", the settings are completely different.

**Comment by yorkshire-pudding** [comment]
There are currently two ways to set the field to have no limit:
![image](https://user-images.githubusercontent.com/20357797/183498037-50ae9e0f-f3aa-45ad-b41b-9e29a1b686af.png)
I think radio button is clearer, so that if the radio buttons are set to "By year" then there must be entries in the Starting year and Ending year boxes.

**Comment by docwilmot** [comment]
I'm wondering though if we shoud just get in the HTML5 widget and then create an issue to add better date limit support to this and all other widgets.

**Comment by indigoxela** [comment]
> I think radio button is clearer, so that if the radio buttons are set to "By year" then there must be entries in the Starting year and Ending year boxes.

@yorkshire-pudding I totally agree that the radio buttons are more intuitive, but IMO there don't have to be entries in *both* number fields. Use case: if one wants a limit for start, but none for end (or the other way round). We could add initial values, though - should we, and which ones? Suggestions?

> I'm wondering though if we shoud just get in the HTML5 widget and then create an issue to add better date limit support to this and all other widgets.

@docwilmot that's exactly what I was thinking. Add the new widget with the "traditional" year limit and in a follow-up add more options - based on the most common use cases (today / relative / fixed years...) TBD.

**Comment by yorkshire-pudding** [comment]
@indigoxela - good point.
I agree that the widget is the most important.  The current year limits default to 3 years either way with relatively sensible options that cover plenty of use cases.  I think the main use case for going backwards many years is date of birth though there will also be a number of recent (post 1582 - year Gregorian calendar introduced) historical events, but the current year setting fields allow that as you can select other and give a greater number.

**Comment by indigoxela** [comment]
> The current year limits default to 3 years either way...

Fair enough :+1: Let's stay in line with existing widget defaults. PR updated.

**Comment by yorkshire-pudding** [comment]
Look good. One wording could be improved:
![image](https://user-images.githubusercontent.com/20357797/183623762-a83f8a21-810b-4bb3-b3f3-311e477ea354.png)

Suggested change from:
```
Applies to start and end if enabled, is relative to current date.
```
To:
```
Applies to stand and end if enabled and is relative to the current date.
```

Typo in:
```
If using simple inputs such a a single textfield or the date popup, omitting the fieldset may provide simpler styling.
```
Should be "such as a single" 
Realised that the typo is existing core rather than your PR.  Don't know if that needs a separate ticket or not.

**Comment by indigoxela** [comment]
> Applies to stand and end

Are you sure? Which "stand"?

> If using simple inputs such a a single...

Nice catch, I blindly copy/pasted. :smile:  Yes, that an ideal "good first issue": #5713 
But we should also fix it here.

**Comment by yorkshire-pudding** [comment]
😂  A typo of my own - corrected now

**Comment by indigoxela** [comment]
Does anyone here want to set the milestone?

**Comment by yorkshire-pudding** [comment]
@indigoxela - I presume because its UX, it can go into a bug release so added to 1.22.3

**Comment by indigoxela** [comment]
@yorkshire-pudding it's marked as "feature request", so it has to go to the next minor release - 1.23.

**Comment by indigoxela** [comment]
Functional tests now also run for the new widget.

**Comment by yorkshire-pudding** [comment]
I was checking validation with multiple date values and if the end date is before the start date on the second or later value, then when validation happens, it is the first value that is highlighted.
![image](https://user-images.githubusercontent.com/20357797/184221023-f55feec6-5330-472b-8475-fa7c5e67b2f6.png)
I checked on pop-up calendar and this identifies which one is wrong, both in the message 
![image](https://user-images.githubusercontent.com/20357797/184221172-7004ca7d-3b68-4cbf-b0e8-3b088528686d.png)
 and in the highlighting
![image](https://user-images.githubusercontent.com/20357797/184221458-7e46a319-4095-4a1c-8e7b-7a9db7ea726f.png)

**Comment by indigoxela** [comment]
Right, the item delta wasn't correctly considered. PR has been updated. All errors now contain field label and - if it's a multi-value field - the field delta.

**Comment by indigoxela** [comment]
We have several topics that should go to follow-up issues:

- Add alter hooks if necessary (considering contrib "Date all day")
- Extended date limit settings (like identified in previous comments)
- Views exposed form support

**Comment by yorkshire-pudding** [comment]
> Right, the item delta wasn't correctly considered. PR has been updated. All errors now contain field label and - if it's a multi-value field - the field delta.

Thanks @indigoxela - I've tested and that works great now.

> We have several topics that should go to follow-up issues:
> 
> * Add alter hooks if necessary (considering contrib "Date all day")
> * Extended date limit settings (like identified in previous comments)
> * Views exposed form support

I agree that these should be follow up issues.

**Comment by yorkshire-pudding** [comment]
As I mentioned in https://github.com/backdrop/backdrop/pull/4157#pullrequestreview-1072376055 - I've been through this and not spotted anything wrong.  My testing all now works.
@indigoxela - are you happy with my code review (with caveat that I didn't understand everything) or do you need another code review? If so, I'll remove the `pr - needs code review` label
I think this should be ready to be included in 1.23.0

**Comment by indigoxela** [comment]
@yorkshire-pudding I'm more than happy with and thankful for all your thorough testing and feedback. :+1:

Re code (implementation details...) I'd like to have some additional feedback by core experts. Maybe @herbdool finds the time?
And maybe @docwilmot and @bugfolder who already showed some interest here, want to share their thoughts?

**Comment by klonos** [comment]
@indigoxela thank you so much for working on this! 🙏🏼 🙏🏼 🙏🏼 

Here's a few thoughts for improvement:

- Why not use [`type="datetime-local"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local) for the widget if the granularity has been set to more than just year+month+day? This would result in a "unified" date+time field instead of two separate fields for date and time:
  ![image](https://user-images.githubusercontent.com/2423362/185754773-01aa7355-58af-4401-b21a-a799dd4f4502.png)
- With regards to the start/end year limits, since I am also not a big fan of "Leave this field empty to do xyz", how about something like this instead?:
  ![image](https://user-images.githubusercontent.com/2423362/185758171-b41e6d4e-7002-416b-b6a5-63a8aa68ca5a.png)
  ...notice how there's no need for any help text, because everything is self-explanatory via the checkbox labels and the field prefixes/suffixes.

PS: a very commonly-requested feature is to have the time picker be in increments of 5min (or also 15min/30min), but it seems that all browsers have dropped support for that. Even if you specify `step="300"` in the input, the widget still shows increments of 1min - it's just that the validation when you submit the form complains. Hating that.

**Comment by indigoxela** [comment]
> Why not use [type="datetime-local"](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local) for the widget if the granularity

That's not available via core FAPI. And that input type (still) has [incomplete browser support](https://caniuse.com/?search=datetime-local). Many just show a plain text field for input - which isn't comfortable at all.

Many thanks for the text suggestions for year limit input. :+1: 

The step attribute is unreliable - some browsers on some OSes support it the way it's described, but by far not all. I wouldn't recommend to do anything beyond what we already do with it (hide the seconds input), because that would require to add polyfills (and testing that on all platforms / browsers).

**Comment by klonos** [comment]
> And that input type (still) has [incomplete browser support](https://caniuse.com/?search=datetime-local).

Well, only IE11 and below really ...but you are right 👍🏼 ...I guess we'll need to wait till after #214 (and even longer than that) 🤷🏼 

>The step attribute is unreliable...

Agreed 👍🏼 ...I just mentioned it because I find it odd that this is not supported by browser HTML5 widgets, while it seems to be a very commonly-requested feature.

**Comment by indigoxela** [comment]
> Well, only IE11 and below really...

Nope, also Firefox ESR, for example. @klonos Usually you rely on Drupal 9+ - which also avoids to use that type and uses date and time combined (like we also do in our FAPI). Changing the FAPI element (which we re-use here) is not the topic of this issue.

Re "step" and browser support - well, :shrug: not our task. Support is unreliable, so let's not use it in a way that won't work, anyway.

**Comment by yorkshire-pudding** [comment]
Welcome back to the issue queues @klonos - I like the improved year limit UX - great idea.
As @indigoxela says, we need to keep the scope focussed so we can get this in.  Some issues are wider than just this widget so out of scope and some are follow up issues that shouldn't block getting this in.

**Comment by klonos** [comment]
> Welcome back to the issue queues @klonos

Good to be back 🙂 ...still lots to catch up with.

Yes, agreed re not scope-creeping with the FAPI elements 👍🏼

**Comment by indigoxela** [comment]
@klonos as I renamed everything as requested, I also wanted to update the settings form text, but...

- Note that we have radios to switch the limit type. That's to prepare for the additional limit types in a follow-up. Note also that the radios are more intuitive than what the date module currently does (has been confirmed here)
- Note also that the suffix text wouldn't work, as the numbers are negative/positive. That's also to prepare for later tweaks.

So I didn't change the text yet.

@yorkshire-pudding and other testers:

**Important for local testing:** if you have fields with the h5widget, switch to another one before you update the PR code. The whole namespace has changed! Alternatively start with a fresh install.

**Comment by yorkshire-pudding** [comment]
Just done a quick test on tugboat (will aim to do some more extensive testing later) - default value for end date is not populating with any of the options

Otherwise, looks good and the limit type switch is clearer

**Comment by indigoxela** [comment]
> ... default value for end date is not populating

Wow, another good catch!

**Comment by indigoxela** [comment]
PR has been updated to fix value2 defaults (and a new issue has been opened for a flaw found while testing #5744).

**Comment by laryn** [comment]
This looks really good! My only question while testing was the label `Limit base` which seemed a bit cryptic.

**Comment by herbdool** [comment]
I agree about `Limit base`. Perhaps `Limit by` or `Limit type` or just `Limit`. (I'm leaning towards the last option).

I also think the fieldsets should match the other date widgets for consistency. "Date input limits" would be "Date entry" (and collapsible). "Default values" fieldset would be below "Date entry". That makes switching between the widgets more consistent and means any UI improvements could be applied to all more easily.

I think these are all tiny changes so think it won't be Needs work for long.

**Comment by indigoxela** [comment]
I agree that the wording can be improved. :+1: 

> I also think the fieldsets should match the other date widgets for consistency.

I _would_ agree, if the current widget settings form was good, but it isn't. As this is a new widget, this is THE chance to get rid of some of those odd an confusing items, that just bloat the form without providing real benefit.

> any UI improvements could be applied to all more easily

We struggled several times to improve the date field and widget settings forms, but didn't get far. My conclusion after struggling hard with the existing code: start fresh and do it better this time. I don't think that all date field widgets have to be the same (and equally odd).

**Comment by herbdool** [comment]
Small UX improvements can be made in any patch release so this can be changed at anytime if we'd like to make improvements to all the date widget fieldsets. And unlike the two examples you gave @indigoxela, the changes to the fieldsets are dead simple.

**Comment by indigoxela** [comment]
OK, to not derail this with an opinionated discussion about what is "dead simple" or not... back to topic.

@herbdool you seem to object to something, but I'm not sure what - you don't give concrete improvement suggestions (code-wise).

Note: If you insist in using exactly the same odd widget settings form like the other date widgets, we have a problem.

**Comment by herbdool** [comment]
@indigoxela I've posted a review with more specific code changes.

**Comment by stpaultim** [comment]
I took a quick look and appreciate the work that has went into this nice improvement. Not sure I have anything to add at this point.

**Comment by indigoxela** [comment]
> I've posted a review with more specific code changes.

@herbdool many thanks! :pray:  I've addressed your requests (except for the one about the `#fieldset` property - explanation on the PR). Ready for another round of review.

**Comment by herbdool** [comment]
Thank you @indigoxela. Much closer. I responded about the `#fieldset` and investigated some more. I still believe it's needed here and added comments, so put it back to Needs work. Apologies for being a stickler on this one thing.

Perhaps another core committer has a different view and welcome their response here if they want to chime in.

**Comment by indigoxela** [comment]
> I responded about the #fieldset...

Yeah, many thanks for your explanations, I responded again.

> Apologies for being a stickler on this one thing.

Sorry for being such a stubborn developer :smile:, but I'm convinced that your conclusions about that fieldset aren't correct. I provided some screenshots in my PR comments to clarify what I mean.

**Comment by herbdool** [comment]
I made a PR https://github.com/indigoxela/backdrop/pull/5 against yours, @indigoxela, after looking more closely. So let's leave off the `'#fieldset' => 'date_format'` I was suggesting (even though I think it's the best approach), but I've got some other improvements that I've discovered after reviewing it some more:

- add a `#weight` to the elements so that "Default values" fieldset sinks to the bottom. It's important that it matches the default UX layout of all other core fields. If we don't use `'#fieldset' => 'date_format'`, which results in the same re-ordering, then this is the next best thing.
- moved `_date_html5_field_widget_settings_form` to `date.admin.inc` which is also where `_date_field_widget_settings_form` already is.
- added ability to alter `_date_html5_field_widget_settings_form` so it matches `_date_field_widget_settings_form`.
- Made `date_field_widget_form` clearer. Instead of plopping an `if and return` at the beginning, which makes it harder to understand the flow, I made it into a `switch`,  and now only has one return at the end. I also moved the building of the `$element` into helper functions. That's less important but does make it easier to read too.
- I tried to figure out why `!isset($form['#entity']` is needed when the other date widgets don't require it. I'm still not sure why and it's kinda annoying.

My repo is a little bit ahead of yours so my PR included some recent commits that your PR does not. If you do a `git pull --rebase upstream 1.x` or whatever you call your remote branch then I think they'll disappear from my PR.

**Comment by quicksketch** [comment]
I read over the current PR and the main thing I found was also covered by @herbdool: Split `date_field_widget_form()` into separate functions, which is the same thing that is done for other widget hooks. :+1:

**Comment by indigoxela** [comment]
@herbdool I rebased my PR, but your PR has conflicts. Please fix.

**Comment by herbdool** [comment]
@indigoxela I resolved the conflicts.

**Comment by indigoxela** [comment]
Merged. (Touching existing code was actually something I tried to _avoid_, but sure, why not... Who am I to argue...)

I belief, this now needs testing again, because of the code restructure.

**Comment by indigoxela** [comment]
Whoopsie...
```
Fatal error: Cannot redeclare _date_html5_field_widget_settings_form()
```
@herbdool as this came in with your PR against my PR, please fix.

**Comment by herbdool** [comment]
@indigoxela I edited your PR

**Comment by indigoxela** [comment]
@herbdool many thanks! It works again, and tests are also passing again. :+1:

**Comment by quicksketch** [comment]
I'm mostly leaning on @herbdool and @indigoxela's feedback cycle here to have caught anything significant. I re-reviewed the overall structure and it looks much better now with the separate sub-functions for each widget. Test coverage includes the new HTML5 widget running the same tests as other widgets, so I think we should be good.

I merged https://github.com/backdrop/backdrop/pull/4157 into 1.x for 1.23.0! Great teamwork pulling this together with rapid reviews and incorporating feedback here at the end of the release cycle. Extra thanks to @indigoxela, @herbdool, and @yorkshire-pudding who have worked on this in earnest for the last month!

https://github.com/backdrop/backdrop/commit/e77507c358731ce62f336ce8e89f5726fc53cd08 by @indigoxela, @herbdool, @yorkshire-pudding, @klonos, @laryn, @bugfolder, @BWPanda, @quicksketch, and @docwilmot.

Great job team!
